### ISSUE TITLE
How can we speed up Yolov5??

### ISSUE THREAD

**Issue by VyasVedant** [ultralytics/yolov5#8065]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Can anyone suggest how can I speed up program since I am using cpu and no gpu, it takes lot of time so are there any other module which can help code to speed up @glenn-jocher ???

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about **inference speed issues**. PyTorch Hub speeds will vary by hardware, software, model, inference settings, etc. Our default example in Colab with a V100 looks like this:

<img width="1462" alt="Screen Shot 2022-05-03 at 10 20 39 AM" src="https://user-images.githubusercontent.com/26833433/166505744-b7fe9b9a-d2b3-4e95-9441-fa540e0804ef.png">


YOLOv5 🚀 can be run on CPU (i.e. `--device cpu`, slow) or GPU if available (i.e. `--device 0`, faster). You can determine your inference device by viewing the YOLOv5 console output:

### detect.py inference
```bash
python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images/
```

<img width="1039" alt="Screen Shot 2022-05-03 at 2 48 42 PM" src="https://user-images.githubusercontent.com/26833433/166572240-ae598b34-1084-40f0-a19b-bf00fd3f9050.png">


### YOLOv5 PyTorch [Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) inference
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()
# Speed: 631.5ms pre-process, 19.2ms inference, 1.6ms NMS per image at shape (2, 3, 640, 640)
```

### Increase Speeds

If you would like to increase your inference speed some options are:

- Use batched inference with [YOLOv5 PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
- Reduce `--img-size`, i.e. 1280 -> 640 -> 320
- Reduce model size, i.e. YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s -> YOLOv5n
- Use half precision FP16 inference with `python detect.py --half` and `python val.py --half`
- Use a faster GPUs, i.e.: P100 -> V100 -> A100
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to ONNX or OpenVINO for up to 3x CPU speedup ([CPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6613))
- [Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) to TensorRT for up to 5x GPU speedup ([GPU Benchmarks](https://github.com/ultralytics/yolov5/pull/6963))
- Use a free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Good luck 🍀 and let us know if you have any other questions!

**Comment by dnth** [comment]
@VyasVedant you can also consider using DeepSparse to significantly speedup the inference speed. I wrote a tutorial on my blog on how I got 180+ FPS inference on a CPU using only 4 cores.

Hope you'll find it useful 👇

https://dicksonneoh.com/portfolio/supercharging_yolov5_180_fps_cpu/

I also wrote a tutorial on OpenVINO 👇

https://dicksonneoh.com/portfolio/how_to_10x_your_od_model_and_deploy_50fps_cpu/

**Comment by marcortiz11** [comment]
I suggest you to take a look at [sparseml](https://github.com/neuralmagic/sparseml). They provide tools to quantize and prune the yolov5 with minimal prediction degradation. After that, you can deploy your model and run it with the deepsparse engine. It's an engine they developed to run sparse models on CPU. This can make your yolov5 run  pretty fast with speeds comparable to GPU's.

**Comment by AbdelsalamHaa** [comment]
You can speed up yolov5 2x on CPU by quantizing the model using onnxruntime. This will work for raspberrypi4 as well. You can check out this blog on how to achieve that.

https://medium.com/@abdelsalam.h.a.a/boosting-yolov5-performance-on-cpu-with-quantization-techniques-for-raspberry-pi4-too-dc2e24f68269

**Comment by apanand14** [comment]
> @VyasVedant you can also consider using DeepSparse to significantly speedup the inference speed. I wrote a tutorial on my blog on how I got 180+ FPS inference on a CPU using only 4 cores.
> 
> Hope you'll find it useful 👇
> 
> https://dicksonneoh.com/portfolio/supercharging_yolov5_180_fps_cpu/
> 
> I also wrote a tutorial on OpenVINO 👇
> 
> https://dicksonneoh.com/portfolio/how_to_10x_your_od_model_and_deploy_50fps_cpu/

@dnth I would like to deploy yolov8-seg model using deepsparse for the faster inference. Is it possible or it doesn't support yet? Thank you for your answer in advance.

**Comment by glenn-jocher** [comment]
Hello @apanand14,

While DeepSparse can significantly speed up inference on CPU, it is important to note that it is designed specifically for running sparse models. As of my knowledge, DeepSparse does not currently support the deployment of the yolov8-seg model.

If you are looking for ways to optimize inference speed with YOLOv5, I suggest considering some of the other techniques mentioned in this thread, such as using quantization techniques with onnxruntime or exploring the use of OpenVINO.

Thank you for your question and if you have any further issues or questions, feel free to ask.

Glenn

**Comment by JyothiRaparla** [comment]
Hello. I am using YOLO11 and converting the it to ONNX model to integrate with my winform app for video analysis of 5 minutes length videos to detect objects. But it is taking 20minutes time to complete the process. Could you suggest me how can I reduce that time

**Comment by pderrenger** [comment]
Hello, for improving inference speed when using ONNX models with your WINFORM app, consider the following:

1. **Optimize the ONNX model** using quantization techniques with tools like `onnxruntime` or frameworks like Neural Magic’s DeepSparse for sparsified models. DeepSparse can significantly speed up inference, achieving GPU-level performance on CPUs: [Neural Magic DeepSparse Integration](https://docs.ultralytics.com/integrations/neural-magic/).

2. **Use smaller YOLO models:** Try using smaller, more efficient model versions like YOLOv5s or YOLOv5n, as they require less computation while maintaining accuracy.

3. **Optimize input resolution:** Running the model on reduced input image sizes (e.g., 640x640 instead of higher resolutions) reduces processing time.

4. **Batch process frames** to reduce repetitive overhead during inference.

If you're not currently leveraging these optimizations, they could provide substantial time reductions for your task. Let us know if you need further guidance!
